### PR TITLE
Speed up image preprocessing for vision-language models

### PR DESCRIPTION
**Pull request by labAxiaoming**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47453)

Avoid materializing repeated temporal copies when packing static image patches in GLM4V, GLM46V, GLMGA, MiniMaxM3-VL, and PaddleOCR-VL.

Skip the final concatenation for single-image inputs across related image processors, and regenerate files derived from modular sources.

The optimized paths preserve the existing patch order and produce bitwise-identical outputs.

# What does this PR do?

This PR speeds up Torchvision image preprocessing by avoiding redundant tensor
copies in several vision-language image processors.

This is a follow-up to #45719, which optimized `Qwen2VLImageProcessor` and the
modular models derived from it at that time. Several independent or newer
processors still used the previous static-image patch packing path:

```python
patches = patches.unsqueeze(1)
repeats = patches[:, -1:].repeat(...)
patches = torch.cat([patches, repeats], dim=1)
```

For static images, every temporal patch contains repeated copies of the same
image. The new implementation first packs the spatial patches, then introduces
the temporal dimension with `expand` before the final `reshape`:

```python
patches = patches.permute(...)
flatten_patches = (
    patches.unsqueeze(6)
    .expand(-1, -1, -1, -1, -1, -1, temporal_patch_size, -1, -1)
    .reshape(batch_size, grid_h * grid_w, channel * temporal_patch_size * patch_size * patch_size)
)
```

This removes the full-size `repeat` intermediate and its following `cat` while
preserving the exact patch order. PaddleOCR-VL uses
`temporal_patch_size = 1`, so its Torchvision path can pack the spatial patches
directly without introducing a temporal dimension.

The PR also avoids `torch.cat` when preprocessing a single image:

```python
pixel_values = processed_images[0] if len(processed_images) == 1 else torch.cat(processed_images, dim=0)
```

The patch packing optimization applies to:

- GLM4V
- GLM46V
- GLMGA
- MiniMaxM3-VL
- PaddleOCR-VL

The single-image `cat` shortcut is also applied to the affected Qwen2-VL-style
and related processors, including Ernie4.5-VL-MoE, GLM Image, HunyuanVL,
Kimi K2.5, Qwen2-VL, and VideoLLaMA3.

Modular sources were updated where applicable and the generated files were
regenerated.

## Benchmark

The benchmark compares the optimized processor imported from the current
checkout and working tree against the previous commit. The baseline processor
source is loaded with `git show HEAD~1:<file>`, while the optimized processor is
imported normally from the current source tree. Both runs use seeded
`1024 x 1024` PIL images, one Torch CPU thread, 5 warmup iterations, and 30
measured iterations. The tables report median latency.

[bench_image_preprocess_optimization.py](https://github.com/user-attachments/files/30225688/bench_image_preprocess_optimization.py)

```bash
python bench_image_preprocess_optimization.py --processors all --num-images 2
python bench_image_preprocess_optimization.py --processors all --num-images 1
```

### Two images

With two images, the final `torch.cat` is still required. This workload
therefore isolates the temporal patch packing optimization:

| Processor | Baseline | This PR | Latency reduction |
|---|---:|---:|---:|
| GLM4V | 144.71 ms | 109.26 ms | 24.5% |
| GLM46V | 130.03 ms | 98.44 ms | 24.3% |
| GLMGA | 131.26 ms | 104.33 ms | 20.5% |
| MiniMaxM3-VL | 66.51 ms | 51.48 ms | 22.6% |
| PaddleOCR-VL | 60.83 ms | 60.87 ms | -0.1% |

PaddleOCR-VL has `temporal_patch_size = 1`, so the removed temporal operations
do not materialize repeated image data. Its two-image performance is therefore
unchanged. Ernie4.5-VL-MoE, GLM Image, HunyuanVL, Kimi K2.5, Qwen2-VL, and
VideoLLaMA3 only receive the single-image `cat` shortcut. With two images they
take the same `torch.cat` branch as the baseline, so any measured difference is
benchmark noise rather than an expected speedup.

### One image

With one image, the final concatenation is skipped. The GLM-family and
MiniMaxM3-VL results include both the patch packing optimization and the
single-image shortcut:

| Processor | Baseline | This PR | Latency reduction |
|---|---:|---:|---:|
| Ernie4.5-VL-MoE | 38.26 ms | 30.74 ms | 19.7% |
| GLM4V | 62.21 ms | 37.91 ms | 39.1% |
| GLM46V | 62.83 ms | 39.17 ms | 37.7% |
| GLM Image | 12.68 ms | 10.08 ms | 20.5% |
| GLMGA | 64.30 ms | 37.01 ms | 42.4% |
| HunyuanVL | 21.80 ms | 17.51 ms | 19.7% |
| Kimi K2.5 | 21.78 ms | 14.93 ms | 31.5% |
| MiniMaxM3-VL | 31.21 ms | 18.84 ms | 39.7% |
| PaddleOCR-VL | 36.07 ms | 28.82 ms | 20.1% |
| Qwen2-VL | 47.77 ms | 36.08 ms | 24.5% |
| VideoLLaMA3 | 33.80 ms | 27.77 ms | 17.8% |

For every benchmarked processor, the baseline and current outputs have
identical keys, shapes, and dtypes, and pass strict `torch.equal` checks:

```text
max_abs_diff={'pixel_values': 0.0, 'image_grid_thw': 0.0}
```

Both the single-image and two-image runs pass the strict comparison across all
affected processors. `image_merge_sizes` is also identical for VideoLLaMA3.

## Tests

- Verified that all affected generated image processor files exactly match the
  modular converter output after Ruff formatting.
- Ran the following targeted image processor tests for GLM4V, HunyuanVL,
  Kimi K2.5, PaddleOCR-VL, Qwen2-VL, and VideoLLaMA3:

```bash
python -m pytest -q -p no:network_debug \
  -k 'not backends_equivalence' \
  tests/models/glm4v/test_image_processing_glm4v.py \
  tests/models/hunyuan_vl/test_image_processing_hunyuan_vl.py \
  tests/models/kimi_k25/test_image_processing_kimi_k25.py \
  tests/models/paddleocr_vl/test_image_processing_paddleocr_vl.py \
  tests/models/qwen2_vl/test_image_processing_qwen2_vl.py \
  tests/models/video_llama_3/test_image_processing_video_llama_3.py
```

```text
117 passed, 31 skipped, 12 deselected
```

This is a targeted test result rather than the full Transformers test suite.
The 12 deselected tests are the backend-equivalence cases excluded by the
`-k` expression.

- Ran formatting and consistency checks:

```bash
git diff --name-only HEAD~1 HEAD -- '*.py' | xargs ruff check
git diff --name-only HEAD~1 HEAD -- '*.py' | xargs ruff format --check
git diff --check HEAD~1 HEAD
```

These commands assume that the optimization is the current `HEAD` commit. Ruff
reported no issues in its 15 modified Python files. The repository-wide
unscoped `ruff check` currently reports pre-existing issues in unrelated files
and was not used as the result above.

## Code Agent Policy

- [ ] I confirm that this is not a pure code agent PR.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [ ] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Models:
- multimodal models: @zucchini-nlp

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @zucchini-nlp

**Comment by labAxiaoming** [comment]
Hello~ Could you please help review this optimization when you have a moment? Thankyou~ @zucchini-nlp

**Comment by zucchini-nlp** [comment]
sorry for delay, reviewing today!

File: src/transformers/models/paddleocr_vl/image_processing_paddleocr_vl.py [hunk 1]
```diff
@@ -188,35 +188,19 @@ def _preprocess(
...
-                grid_h,
-                patch_size,
-                grid_w,
-                patch_size,
-            )
-            patches = patches.permute(0, 1, 4, 6, 3, 2, 5, 7)
-            flatten_patches = patches.reshape(batch_size, grid_t * grid_h * grid_w, channel, patch_size, patch_size)
+            patches = patches.reshape(batch_size, channel, grid_h, patch_size, grid_w, patch_size)
+            patches = patches.permute(0, 2, 4, 1, 3, 5)
+            flatten_patches = patches.reshape(batch_size, grid_h * grid_w, channel, patch_size, patch_size)
```
Comment (zucchini-nlp) [inline src/transformers/models/paddleocr_vl/image_processing_paddleocr_vl.py]: iIrc we don't expand because the temporal patch size is `1`, but let's keep it true to main branch and have an `expand(temporal_patch_size)`
never can know if someone is using the same model with different temporal patch :)

**Review by zucchini-nlp — COMMENTED** [review]
Thanks, lets make PaddleOCR completely BC and we can merge

File: src/transformers/models/paddleocr_vl/image_processing_paddleocr_vl.py (hunk 1 above)
Comment (labAxiaoming) [reply src/transformers/models/paddleocr_vl/image_processing_paddleocr_vl.py]: yes, you are right
fix:
```diff
diff --git a/src/transformers/models/paddleocr_vl/image_processing_paddleocr_vl.py b/src/transformers/models/paddleocr_vl/image_processing_paddleocr_vl.py
--- a/src/transformers/models/paddleocr_vl/image_processing_paddleocr_vl.py
+++ b/src/transformers/models/paddleocr_vl/image_processing_paddleocr_vl.py
@@ -194,7 +194,17 @@ class PaddleOCRVLImageProcessor(TorchvisionBackend):
 
             patches = patches.reshape(batch_size, channel, grid_h, patch_size, grid_w, patch_size)
             patches = patches.permute(0, 2, 4, 1, 3, 5)
-            flatten_patches = patches.reshape(batch_size, grid_h * grid_w, channel, patch_size, patch_size)
+            flatten_patches = (
+                patches.unsqueeze(4)
+                .expand(-1, -1, -1, -1, temporal_patch_size, -1, -1)
+                .reshape(
+                    batch_size,
+                    grid_h * grid_w,
+                    channel * temporal_patch_size,
+                    patch_size,
+                    patch_size,
+                )
+            )
 
             processed_images_grouped[shape] = flatten_patches
             processed_grids[shape] = [[1, grid_h, grid_w]] * batch_size
diff --git a/src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py b/src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py
--- a/src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py
+++ b/src/transformers/models/paddleocr_vl/modular_paddleocr_vl.py
@@ -306,7 +306,17 @@ class PaddleOCRVLImageProcessor(Qwen2VLImageProcessor):
 
             patches = patches.reshape(batch_size, channel, grid_h, patch_size, grid_w, patch_size)
             patches = patches.permute(0, 2, 4, 1, 3, 5)
-            flatten_patches = patches.reshape(batch_size, grid_h * grid_w, channel, patch_size, patch_size)
+            flatten_patches = (
+                patches.unsqueeze(4)
+                .expand(-1, -1, -1, -1, temporal_patch_size, -1, -1)
+                .reshape(
+                    batch_size,
+                    grid_h * grid_w,
+                    channel * temporal_patch_size,
+                    patch_size,
+                    patch_size,
+                )
+            )
 
             processed_images_grouped[shape] = flatten_patches
             processed_grids[shape] = [[1, grid_h, grid_w]] * batch_size


```

**Comment by zucchini-nlp** [comment]
@bot /repo

**Review by zucchini-nlp — APPROVED** [review]
ah right, needs an approve first. Thanks
